File: app/src/main/java/com/sortedqueue/learnandroid/topic/CodeFragment.java
package com.sortedqueue.learnandroid.topic;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.pddstudio.highlightjs.HighlightJsView;
import com.pddstudio.highlightjs.models.Language;
import com.pddstudio.highlightjs.models.Theme;
import com.sortedqueue.learnandroid.R;
import com.sortedqueue.learnandroid.asynctasks.CodeFileReaderTask;
import com.sortedqueue.learnandroid.dashboard.DashboardNavigationListener;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;


/**
 * Created by Alok on 06/06/17.
 */

public class CodeFragment extends Fragment implements CodeFileReaderTask.OnDataReadListener, HighlightJsView.OnThemeChangedListener {

    @BindView(R.id.highlightJsView)
    HighlightJsView highlightJsView;
    Unbinder unbinder;
    @BindView(R.id.backTextView)
    TextView backTextView;
    @BindView(R.id.codeProgressBar)
    ProgressBar codeProgressBar;
    private SlideContent slideContent;

    private DashboardNavigationListener dashboardNavigationListener;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View fragmentView = inflater.inflate(R.layout.fragment_code, container, false);
        unbinder = ButterKnife.bind(this, fragmentView);
        //register theme change listener
        highlightJsView.setOnThemeChangedListener(this);
        //change theme and set language to auto detect
        highlightJsView.setTheme(Theme.ANDROID_STUDIO);
        String codeFile = slideContent.getContent();
        if (codeFile.startsWith("xml")) {
            highlightJsView.setHighlightLanguage(Language.XML);
        } else {
            highlightJsView.setHighlightLanguage(Language.JAVA);
        }
        //load the source (can be loaded by String, File or URL)
        highlightJsView.setShowLineNumbers(true);
        highlightJsView.setZoomSupportEnabled(true);
        highlightJsView.getViewTreeObserver().addOnScrollChangedListener(new ViewTreeObserver.OnScrollChangedListener() {
            @Override
            public void onScrollChanged() {
                if( highlightJsView == null ) {
                    return;
                }
                View view = highlightJsView.getChildAt(highlightJsView.getChildCount() - 1);
                if( view != null ) {
                    int diff = (view.getBottom() - (highlightJsView.getHeight() + highlightJsView
                            .getScrollY()));

                    if (diff == 0) {
                        presentationSlideListener.showNextLayout();
                    }
                    else {
                        presentationSlideListener.hideNextLayout();
                    }
                }
                else {
                    presentationSlideListener.showNextLayout();
                }

            }
        });
        backTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dashboardNavigationListener.onNavigateBack();
            }
        });
        new CodeFileReaderTask(codeProgressBar, getContext(), slideContent.getContent(), this).execute();
        return fragmentView;
    }

    private PresentationSlideListener presentationSlideListener;

    public void setPresentationSlideListener(PresentationSlideListener presentationSlideListener) {
        this.presentationSlideListener = presentationSlideListener;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof DashboardNavigationListener) {
            dashboardNavigationListener = (DashboardNavigationListener) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        dashboardNavigationListener = null;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDataReadComplete(String code) {
        highlightJsView.setSource(code);
        codeProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void onThemeChanged(@NonNull Theme theme) {

    }

    public void setSlideContent(SlideContent slideContent) {
        this.slideContent = slideContent;
    }
}

File: app/src/main/java/com/sortedqueue/learnandroid/topic/CodeFragment.kt
package com.sortedqueue.learnandroid.topic

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver

import com.pddstudio.highlightjs.HighlightJsView
import com.pddstudio.highlightjs.models.Language
import com.pddstudio.highlightjs.models.Theme
import com.sortedqueue.learnandroid.R
import com.sortedqueue.learnandroid.asynctasks.CodeFileReaderTask
import com.sortedqueue.learnandroid.dashboard.DashboardNavigationListener
import kotlinx.android.synthetic.main.fragment_code.*

/**
 * Created by Alok on 06/06/17.
 */

class CodeFragment : Fragment(), CodeFileReaderTask.OnDataReadListener, HighlightJsView.OnThemeChangedListener {

    private var slideContent: SlideContent? = null

    private var dashboardNavigationListener: DashboardNavigationListener? = null

    private var presentationSlideListener: PresentationSlideListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_code, container, false)
    }

    /**
     * Called immediately after [.onCreateView]
     * has returned, but before any saved state has been restored in to the view.
     * This gives subclasses a chance to initialize themselves once
     * they know their view hierarchy has been completely created.  The fragment's
     * view hierarchy is not however attached to its parent at this point.
     * @param view The View returned by [.onCreateView].
     * @param savedInstanceState If non-null, this fragment is being re-constructed
     * from a previous saved state as given here.
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //register theme change listener
        highlightJsView.setOnThemeChangedListener(this)
        //change theme and set language to auto detect
        highlightJsView.setTheme(Theme.ANDROID_STUDIO)
        val codeFile = slideContent!!.content
        if (codeFile!!.startsWith("xml")) {
            highlightJsView.setHighlightLanguage(Language.XML)
        } else {
            highlightJsView.setHighlightLanguage(Language.JAVA)
        }
        //load the source (can be loaded by String, File or URL)
        highlightJsView.setShowLineNumbers(true)
        highlightJsView.setZoomSupportEnabled(true)
        highlightJsView.getViewTreeObserver().addOnScrollChangedListener(ViewTreeObserver.OnScrollChangedListener {
            if (highlightJsView == null) {
                return@OnScrollChangedListener
            }
            val view = highlightJsView.getChildAt(highlightJsView.getChildCount() - 1)
            if (view != null) {
                val diff = view!!.getBottom() - (highlightJsView.getHeight() + highlightJsView
                        .getScrollY())

                if (diff == 0) {
                    presentationSlideListener!!.showNextLayout()
                } else {
                    presentationSlideListener!!.hideNextLayout()
                }
            } else {
                presentationSlideListener!!.showNextLayout()
            }
        })
        backTextView.setOnClickListener(View.OnClickListener { dashboardNavigationListener!!.onNavigateBack() })
        CodeFileReaderTask(codeProgressBar, context!!, slideContent!!.content!!, this).execute()
    }

    fun setPresentationSlideListener(presentationSlideListener: PresentationSlideListener) {
        this.presentationSlideListener = presentationSlideListener
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is DashboardNavigationListener) {
            dashboardNavigationListener = context
        }
    }

    override fun onDetach() {
        super.onDetach()
        dashboardNavigationListener = null
    }


    override fun onDataReadComplete(code: String?) {
        highlightJsView.setSource(code)
        codeProgressBar.setVisibility(View.GONE)
    }

    override fun onThemeChanged(theme: Theme) {

    }

    fun setSlideContent(slideContent: SlideContent) {
        this.slideContent = slideContent
    }
}
